File: dmlab2d/env_lua_api/properties_test.cc
// Copyright (C) 2019 The DMLab2D Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////
#include "dmlab2d/env_lua_api/properties.h"

#include "absl/strings/string_view.h"
#include "dmlab2d/lua/bind.h"
#include "dmlab2d/lua/call.h"
#include "dmlab2d/lua/lua.h"
#include "dmlab2d/lua/n_results_or_test_util.h"
#include "dmlab2d/lua/push_script.h"
#include "dmlab2d/lua/table_ref.h"
#include "dmlab2d/lua/vm_test_util.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "third_party/rl_api/env_c_api.h"

namespace deepmind::lab2d {
namespace {

using ::deepmind::lab2d::lua::testing::IsOkAndHolds;
using ::testing::Eq;
using ::testing::StrEq;

class PropertiesTest : public lua::testing::TestWithVm {
 protected:
  PropertiesTest() {
    vm()->AddCModuleToSearchers("system.properties",
                                &lua::Bind<Properties::Module>);
  }
};

constexpr char kPropertyApi[] = R"(
local properties = require 'system.properties'

local api = {
    _readWriteNumber = 10,
    _readOnlyString = 'Hello',
    _writeString = 'There',
    _emptyList = {},
}

function api:writeProperty(key, value)
  if key == '' then
    return properties.PERMISSION_DENIED
  end
  if key == 'readWriteNumber' then
    v = tonumber(value)
    if v ~= nil then
      self._readWriteNumber = v
      return properties.SUCCESS
    else
      return properties.INVALID_ARGUMENT
    end
  end
  if key == 'readOnlyString' then
    return properties.PERMISSION_DENIED
  end
  if key == 'writeString' then
    self._writeString = value
    return properties.SUCCESS
  end
  if key == 'emptyList' then
    return properties.PERMISSION_DENIED
  end
  return properties.NOT_FOUND
end

function api:readProperty(key)
  if key == '' then
    return properties.PERMISSION_DENIED
  end
  if key == 'readWriteNumber' then
    return tostring(self._readWriteNumber)
  end
  if key == 'readOnlyString' then
    return self._readOnlyString
  end
  if key == 'writeString' then
    return properties.PERMISSION_DENIED
  end
  if key == 'emptyList' then
    return properties.PERMISSION_DENIED
  end
  return properties.NOT_FOUND
end

function api:listProperty(key, callback)
  if key == '' then
    callback('readWriteNumber', 'rw')
    callback('readOnlyString', 'r')
    callback('writeString', 'w')
    callback('emptyList', 'l')
    return properties.SUCCESS
  end
  if key == 'emptyList' then
    return properties.SUCCESS
  end
  if key == 'readWriteNumber' or key == 'readOnlyString' or
      key == 'writeString' then
    return properties.PERMISSION_DENIED
  end
  return properties.NOT_FOUND
end

return api
)";

TEST_F(PropertiesTest, TestRead) {
  ASSERT_THAT(lua::PushScript(L, kPropertyApi, "kPropertyApi"),
              IsOkAndHolds(1));
  ASSERT_THAT(lua::Call(L, 0), IsOkAndHolds(1));
  lua::TableRef table;
  ASSERT_TRUE(IsFound(Read(L, 1, &table)));
  lua_settop(L, 0);
  Properties properties;
  ASSERT_THAT(properties.BindApi(table), IsOkAndHolds(0));
  const char* value = "";
  EXPECT_THAT(properties.ReadProperty("", &value),
              Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(properties.ReadProperty("emptyList", &value),
              Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(properties.ReadProperty("readWriteNumber", &value),
              Eq(EnvCApi_PropertyResult_Success));
  EXPECT_THAT(value, StrEq("10"));
  EXPECT_THAT(properties.ReadProperty("readOnlyString", &value),
              Eq(EnvCApi_PropertyResult_Success));
  EXPECT_THAT(value, StrEq("Hello"));
  EXPECT_THAT(properties.ReadProperty("writeString", &value),
              Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(properties.ReadProperty("notExist", &value),
              Eq(EnvCApi_PropertyResult_NotFound));
}

TEST_F(PropertiesTest, TestWrite) {
  ASSERT_THAT(lua::PushScript(L, kPropertyApi, "kPropertyApi"),
              IsOkAndHolds(1));
  ASSERT_THAT(lua::Call(L, 0), IsOkAndHolds(1));
  lua::TableRef table;
  ASSERT_TRUE(IsFound(Read(L, 1, &table)));
  lua_settop(L, 0);
  Properties properties;
  ASSERT_THAT(properties.BindApi(table), IsOkAndHolds(0));
  EXPECT_THAT(properties.WriteProperty("", "Hello"),
              Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(properties.WriteProperty("emptyList", "Hello"),
              Eq(EnvCApi_PropertyResult_PermissionDenied));

  EXPECT_THAT(properties.WriteProperty("readWriteNumber", "Blah"),
              Eq(EnvCApi_PropertyResult_InvalidArgument));

  EXPECT_THAT(properties.WriteProperty("readWriteNumber", "10"),
              Eq(EnvCApi_PropertyResult_Success));
  int int_value = 0;
  EXPECT_TRUE(table.LookUp("_readWriteNumber", &int_value));
  EXPECT_THAT(int_value, Eq(10));

  EXPECT_THAT(properties.WriteProperty("readOnlyString", ""),
              Eq(EnvCApi_PropertyResult_PermissionDenied));

  EXPECT_THAT(properties.WriteProperty("writeString", "Anything"),
              Eq(EnvCApi_PropertyResult_Success));

  absl::string_view value;
  EXPECT_TRUE(table.LookUp("_writeString", &value));
  EXPECT_THAT(value, Eq("Anything"));

  EXPECT_THAT(properties.WriteProperty("notExists", "Anything"),
              Eq(EnvCApi_PropertyResult_NotFound));
}

TEST_F(PropertiesTest, TestList) {
  ASSERT_THAT(lua::PushScript(L, kPropertyApi, "kPropertyApi"),
              IsOkAndHolds(1));
  ASSERT_THAT(lua::Call(L, 0), IsOkAndHolds(1));
  lua::TableRef table;
  ASSERT_TRUE(IsFound(Read(L, 1, &table)));
  lua_settop(L, 0);
  Properties properties;
  ASSERT_THAT(properties.BindApi(table), IsOkAndHolds(0));

  struct Results {
    std::vector<std::string> names;
    std::vector<EnvCApi_PropertyAttributes> attributes;
  };

  auto callback = +[](void* userdata, const char* key,
                      EnvCApi_PropertyAttributes attributes) {
    auto* results_ptr = static_cast<Results*>(userdata);
    results_ptr->names.emplace_back(key);
    results_ptr->attributes.emplace_back(attributes);
  };

  Results all;
  EXPECT_THAT(properties.ListProperty(&all, "", callback),
              Eq(EnvCApi_PropertyResult_Success));

  EXPECT_THAT(all.names,
              ::testing::ElementsAre("readWriteNumber", "readOnlyString",
                                     "writeString", "emptyList"));
  EXPECT_THAT(all.attributes,
              ::testing::ElementsAre(EnvCApi_PropertyAttributes_ReadWritable,
                                     EnvCApi_PropertyAttributes_Readable,
                                     EnvCApi_PropertyAttributes_Writable,
                                     EnvCApi_PropertyAttributes_Listable));

  Results empty;
  EXPECT_THAT(properties.ListProperty(&empty, "emptyList", callback),
              Eq(EnvCApi_PropertyResult_Success));

  EXPECT_THAT(empty.names, ::testing::IsEmpty());
  EXPECT_THAT(empty.attributes, ::testing::IsEmpty());

  Results results_denied;
  EXPECT_THAT(
      properties.ListProperty(&results_denied, "readWriteNumber", callback),
      Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(results_denied.names, ::testing::IsEmpty());
  EXPECT_THAT(results_denied.attributes, ::testing::IsEmpty());

  EXPECT_THAT(
      properties.ListProperty(&results_denied, "readOnlyString", callback),
      Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(results_denied.names, ::testing::IsEmpty());
  EXPECT_THAT(results_denied.attributes, ::testing::IsEmpty());

  EXPECT_THAT(properties.ListProperty(&results_denied, "writeString", callback),
              Eq(EnvCApi_PropertyResult_PermissionDenied));
  EXPECT_THAT(results_denied.names, ::testing::IsEmpty());
  EXPECT_THAT(results_denied.attributes, ::testing::IsEmpty());

  EXPECT_THAT(properties.ListProperty(&results_denied, "notExists", callback),
              Eq(EnvCApi_PropertyResult_NotFound));
}

}  // namespace
}  // namespace deepmind::lab2d

File: dmlab2d/lib/game_scripts/common/properties_test.lua
--[[ Copyright (C) 2019 The DMLab2D Authors.

Licensed under the Apache License, Version 2.0 (the 'License');
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an 'AS IS' BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
]]

local properties = require 'common.properties'
local asserts = require 'testing.asserts'
local test_runner = require 'testing.test_runner'

local tests = {}

local function _propertyListKey(api, key)
  local list = {}
  local function gather(key, mode)
    list[key] = mode
  end
  asserts.EQ(properties.RESULT.SUCCESS, api:listProperty(key, gather))
  return list
end

function tests.addReadWriteTableWorks()
  local api = {}
  properties.decorate(api)
  local settings = {
      num = 10,
      text = 'hello',
      subTable = {
          subNum = 5,
          subText = 'world'
      }
  }
  properties.addReadWrite('settingsRw', settings)
  local settingsList = _propertyListKey(api, 'settingsRw')
  asserts.tablesEQ(settingsList, {['settingsRw.num'] = 'rw',
                                  ['settingsRw.text'] = 'rw',
                                  ['settingsRw.subTable'] = 'l'})

  local settingsList = _propertyListKey(api, 'settingsRw.subTable')
  asserts.tablesEQ(settingsList, {['settingsRw.subTable.subNum'] = 'rw',
                                  ['settingsRw.subTable.subText'] = 'rw'})

  asserts.EQ(api:readProperty('settingsRw.num'), '10')
  asserts.EQ(api:readProperty('settingsRw.text'), 'hello')
  asserts.EQ(api:readProperty('settingsRw.subTable.subNum'), '5')
  asserts.EQ(api:readProperty('settingsRw.subTable.subText'), 'world')

  asserts.EQ(api:writeProperty('settingsRw.num', '100'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsRw.text', 'Hello'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsRw.subTable.subNum', '50'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsRw.subTable.subText', 'World'),
             properties.RESULT.SUCCESS)

  asserts.tablesEQ(settings, {
          num = 100,
          text = 'Hello',
          subTable = {
              subNum = 50,
              subText = 'World'
          }
      })
  properties.removeReadWrite('settingsRw')
  asserts.EQ(api:writeProperty('settingsRw.num', '100'),
             properties.RESULT.NOT_FOUND)
end

function tests.addWriteOnlyTableWorks()
  local api = {}
  properties.decorate(api)
  local settings = {
      num = 10,
      text = 'hello',
      subTable = {
          subNum = 5,
          subText = 'world'
      }
  }
  properties.addWriteOnly('settingsW', settings)
  local settingsList = _propertyListKey(api, 'settingsW')
  asserts.tablesEQ(settingsList, {['settingsW.num'] = 'w',
                                  ['settingsW.text'] = 'w',
                                  ['settingsW.subTable'] = 'l'})

  local settingsList = _propertyListKey(api, 'settingsW.subTable')
  asserts.tablesEQ(settingsList, {['settingsW.subTable.subNum'] = 'w',
                                  ['settingsW.subTable.subText'] = 'w'})

  asserts.EQ(api:readProperty('settingsW.num'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:readProperty('settingsW.text'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:readProperty('settingsW.subTable.subNum'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:readProperty('settingsW.subTable.subText'),
             properties.RESULT.NOT_FOUND)

  asserts.EQ(api:writeProperty('settingsW.num', '100'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsW.text', 'Hello'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsW.subTable.subNum', '50'),
             properties.RESULT.SUCCESS)
  asserts.EQ(api:writeProperty('settingsW.subTable.subText', 'World'),
             properties.RESULT.SUCCESS)

  asserts.tablesEQ(settings, {
          num = 100,
          text = 'Hello',
          subTable = {
              subNum = 50,
              subText = 'World'
          }
      })

  properties.removeWriteOnly('settingsW')
  asserts.EQ(api:writeProperty('settingsW.num', '100'),
             properties.RESULT.NOT_FOUND)
end


function tests.addReadTableWorks()
  local api = {}
  properties.decorate(api)
  local settings = {
      num = 10,
      text = 'hello',
      subTable = {
          subNum = 5,
          subText = 'world'
      }
  }
  properties.addReadOnly('settingsR', settings)
  local settingsList = _propertyListKey(api, 'settingsR')
  asserts.tablesEQ(settingsList, {['settingsR.num'] = 'r',
                                  ['settingsR.text'] = 'r',
                                  ['settingsR.subTable'] = 'l'})

  local settingsList = _propertyListKey(api, 'settingsR.subTable')
  asserts.tablesEQ(settingsList, {['settingsR.subTable.subNum'] = 'r',
                                  ['settingsR.subTable.subText'] = 'r'})

  asserts.EQ(api:readProperty('settingsR.num'), '10')
  asserts.EQ(api:readProperty('settingsR.text'), 'hello')
  asserts.EQ(api:readProperty('settingsR.subTable.subNum'), '5')
  asserts.EQ(api:readProperty('settingsR.subTable.subText'), 'world')

  asserts.EQ(api:writeProperty('settingsR.num', '100'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:writeProperty('settingsR.text', 'Hello'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:writeProperty('settingsR.subTable.subNum', '50'),
             properties.RESULT.NOT_FOUND)
  asserts.EQ(api:writeProperty('settingsR.subTable.subText', 'World'),
             properties.RESULT.NOT_FOUND)

  asserts.tablesEQ(settings, {
          num = 10,
          text = 'hello',
          subTable = {
              subNum = 5,
              subText = 'world'
          }
      })
  properties.removeReadOnly('settingsR')
  asserts.EQ(api:readProperty('settingsR.num'), properties.RESULT.NOT_FOUND)
end

function tests.addWriteFunctionWorks()
  local api = {}
  properties.decorate(api)
  local param0
  local param1

  local function myFuncWrite(inParam0, inParam1)
    param0 = inParam0
    param1 = inParam1
    return properties.RESULT.SUCCESS
  end

  properties.addWriteOnly('myFuncWrite', myFuncWrite)

  asserts.EQ(api:writeProperty('myFuncWrite', '100'),
             properties.RESULT.SUCCESS)
  asserts.EQ(param0, '100')

  api:writeProperty('myFuncWrite.param.other', '100')
  asserts.EQ(param0, 'param.other')
  asserts.EQ(param1, '100')
end

function tests.addReadFunctionWorks()
  local api = {}
  properties.decorate(api)
  local param
  local function myFuncRead(inParam)
    param = inParam
    return 'hello'
  end

  properties.addReadOnly('myFuncRead', myFuncRead)
  asserts.EQ(api:readProperty('myFuncRead.10'), 'hello')
  asserts.EQ(param, '10')
  asserts.EQ(api:readProperty('myFuncRead'), 'hello')
  asserts.EQ(param, nil)
end

return test_runner.run(tests)
